driver-core/src/main/com/mongodb/internal/async/client/ClientSessionBinding.java

```diff
@@ -131,15 +128,15 @@ public void getReadConnectionSource(final int minWireVersion, final ReadPreferen
     }
 
     @Override
-    public void release() {
-        wrapped.release();
-        closeSessionIfCountIsZero();
-    }
-
-    private void closeSessionIfCountIsZero() {
-        if (getCount() == 0 && ownsSession) {
```

Comment:
Stop using `getCount` and instead give this class its own reference count so that `release` can properly close the session.